src/Compilation.zig

```diff
@@ -3098,42 +3099,111 @@ pub fn performAllTheWork(
-                    note.* = switch (file.references.items[i]) {
-                        .import => |loc| try Module.ErrorMsg.init(
-                            mod.gpa,
-                            loc,
-                            "imported from package {s}",
-                            .{loc.file_scope.pkg.name},
-                        ),
-                        .root => |pkg| try Module.ErrorMsg.init(
-                            mod.gpa,
-                            .{ .file_scope = file, .parent_decl_node = 0, .lazy = .entire_file },
-                            "root of package {s}",
-                            .{pkg.name},
-                        ),
+                    note.* = switch (ref) {
+                        .import => |loc| blk: {
+                            const name = try loc.file_scope.pkg.getName(mod.gpa, mod.*);
+                            defer mod.gpa.free(name);
+                            break :blk try Module.ErrorMsg.init(
+                                mod.gpa,
+                                loc,
+                                "imported from module {s}",
+                                .{name},
+                            );
+                        },
+                        .root => |pkg| blk: {
+                            const name = try pkg.getName(mod.gpa, mod.*);
+                            defer mod.gpa.free(name);
+                            break :blk try Module.ErrorMsg.init(
+                                mod.gpa,
+                                .{ .file_scope = file, .parent_decl_node = 0, .lazy = .entire_file },
+                                "root of module {s}",
+                                .{name},
+                            );
+                        },
                     };
                 }
-                errdefer for (notes) |*n| n.deinit(mod.gpa);
+                errdefer for (notes[0..num_notes]) |*n| n.deinit(mod.gpa);
+
+                if (omitted > 0) {
+                    notes[num_notes] = try Module.ErrorMsg.init(
+                        mod.gpa,
+                        .{ .file_scope = file, .parent_decl_node = 0, .lazy = .entire_file },
+                        "{} more references omitted",
+                        .{omitted},
+                    );
+                }
+                errdefer if (omitted > 0) notes[num_notes].deinit(mod.gpa);
 
                 const err = try Module.ErrorMsg.create(
                     mod.gpa,
                     .{ .file_scope = file, .parent_decl_node = 0, .lazy = .entire_file },
-                    "file exists in multiple packages",
+                    "file exists in multiple modules",
                     .{},
                 );
                 err.notes = notes;
                 break :err_blk err;
             };
             errdefer err.destroy(mod.gpa);
             try mod.failed_files.putNoClobber(mod.gpa, file, err);
+            last_err = err;
+        }
```

Comment:
This is a rather large percent of `performAllTheWork` that is being used for reporting "file exists in multiple modules". Would you mind extracting this into a separate function?

Reply:
Can do!